controller/src/test/java/io/pravega/controller/task/Stream/StreamMetadataTasksTest.java

```diff
@@ -1020,6 +1021,446 @@ public void sizeBasedRetentionStreamTest() throws Exception {
+        streamMetadataTasks.updateSubscriberStreamCut(SCOPE, stream1, subscriber2, ImmutableMap.of(0L, 11L, 1L, 11L), null).join();
+
+        streamMetadataTasks.retention(SCOPE, stream1, retentionPolicy, time.get(), null, "").join();
+        // now retention set has five stream cuts 1: 0/2, 1/2...10: 0/10, 1/10... 20: 0/11, 1/11.. 30: 0/20, 1/20.. 40L: 0/21, 1/21
+        // subscriber lowerbound is 0/11, 1/11 
+        // max = 30. truncate at max
+        truncationRecord = streamStorePartialMock.getTruncationRecord(SCOPE, stream1, null, executor).join();
+        assertEquals(truncationRecord.getObject().getStreamCut().get(0L).longValue(), 20L);
+        assertEquals(truncationRecord.getObject().getStreamCut().get(1L).longValue(), 20L);
+        assertTrue(truncationRecord.getObject().isUpdating());
+        streamStorePartialMock.completeTruncation(SCOPE, stream1, truncationRecord, null, executor).join();
+        // endregion
+
+        // region case 5: lowerbound overlaps with max
+        streamStorePartialMock.addStreamCutToRetentionSet(SCOPE, stream1,
+                new StreamCutRecord(50L, 43L, ImmutableMap.of(0L, 21L, 1L, 22L)), null, executor).join();
+        time.set(59L);
+        streamStorePartialMock.addStreamCutToRetentionSet(SCOPE, stream1,
+                new StreamCutRecord(time.get(), 60L, ImmutableMap.of(0L, 30L, 1L, 30L)), null, executor).join();
+        time.set(60L);
+        streamStorePartialMock.addStreamCutToRetentionSet(SCOPE, stream1,
+                new StreamCutRecord(time.get(), 60L, ImmutableMap.of(0L, 30L, 1L, 30L)), null, executor).join();
+
+        // update both readers to make sure they have read till the latest position - 1. we have set the min limit to 2.  
+        streamMetadataTasks.updateSubscriberStreamCut(SCOPE, stream1, subscriber1, ImmutableMap.of(0L, 22L, 1L, 21L), null).join();
+        streamMetadataTasks.updateSubscriberStreamCut(SCOPE, stream1, subscriber2, ImmutableMap.of(0L, 22L, 1L, 21L), null).join();
+
+        streamMetadataTasks.retention(SCOPE, stream1, retentionPolicy, time.get(), null, "").join();
+        // now retention set has five stream cuts 1: 0/2, 1/2...10: 0/10, 1/10... 20: 0/11, 1/11.. 30: 0/20, 1/20.. 40L: 0/21, 1/21
+        // 50: 0/21, 1/22 ... 59: 0/30, 1/30.. 60: 0/30, 1/30
+        // subscriber lowerbound is 0/22, 1/21 
+        // this overlaps with max. so truncate at streamcut
+        truncationRecord = streamStorePartialMock.getTruncationRecord(SCOPE, stream1, null, executor).join();
+        assertEquals(truncationRecord.getObject().getStreamCut().get(0L).longValue(), 22L);
+        assertEquals(truncationRecord.getObject().getStreamCut().get(1L).longValue(), 21L);
+        assertTrue(truncationRecord.getObject().isUpdating());
+        streamStorePartialMock.completeTruncation(SCOPE, stream1, truncationRecord, null, executor).join();
+        // endregion
+    }
+
```

Comment:
We need to add tests to cover the following corner cases:
1. CBR Retention Policy does not specify any max/min limit (so they are at their default values)
2. Stream does not have any subscribers (# of subscribers = 0 and/or no subscriber has a valid StreamCut... there have been addSubscriber() calls but no updateSubscriberStreamCut() calls so far)
3. Stream has few subscribers with empty StreamCuts (StreamCuts have not been updated from client so far)

Reply:
1. is handled via the CBR policy constructor itself. if nothing is specified the max is long.max and min is defaulted to 0.

2. and 3. 
if one or more subscribers have not updated their streamcut then they are effectively at the mercy of "min". 
should we wait for all subscribers to have reported at least one streamcut? if that is preferred i will make the change.

Reply:
1. I understand that default max==LONG_MAX and default min==0, I'm suggesting having a test case that invokes CBR Stream truncation without these values being explicitly set, just to cover the scenario where customers may not set these limits. I know it should work, but having a test case to validate would be good.
2 & 3 
> if one or more subscribers have not updated their streamcut then they are effectively at the mercy of "min".
should we wait for all subscribers to have reported at least one streamcut? if that is preferred i will make the change.

Yes this is by design, but need to have tests to cover this, so that we know it works as expected.

Reply:
>should we wait for all subscribers to have reported at least one streamcut? if that is preferred i will make the change.

No we should not wait for any subscriber to update their Stream cut. Truncation can happen with 0 subscribers/valid-streamcuts ....its just that it will happen at the max value instead of the lowerBound.
We just need to make sure these corner cases are tested

Reply:
i will add test to cover Long.Max and 0 as max and min respectively.